xla/service/collective_pipeliner.cc

```diff
@@ -527,6 +530,35 @@ std::optional<std::vector<HloInstruction*>> CollectIndependentOperandChain(
         return !IsLoopIterator(instr, loop_iter) &&
                !loop_invariant_params.count(instr);
       };
+
+  // Check if this is a MoveToDevice custom call that follows a dynamic-slice
+  // This is a special case for backward pipelining with host offloading
+  if (instr->IsCustomCall(memory_annotations::kMoveToDeviceCustomCallTarget)) {
+    std::vector<HloInstruction*> to_check = {instr->mutable_operand(0)};
+    std::set<HloInstruction*> visited;
+
+    while (!to_check.empty()) {
+      HloInstruction* current = to_check.back();
+      to_check.pop_back();
+
+      if (visited.insert(current).second) {
+        if (current->opcode() == HloOpcode::kDynamicSlice) {
+          if (visited_set.insert(current).second) {
+            stack.emplace_back(current, 0);
+          }
+          break;
+        }
+        if (current->opcode() == HloOpcode::kReshape ||
+            current->opcode() == HloOpcode::kBroadcast ||
+            current->opcode() == HloOpcode::kTranspose) {
+          for (HloInstruction* operand : current->operands()) {
+            to_check.push_back(operand);
+          }
+        }
+      }
+    }
+  }
```

Comment:
We want to not hide device specific detail from this pass.
Is it possible to add a field to CollectivePipeline::Config, something similar this existing field "HloPredicate acceptable_formatting"?